packages/test/src/test-integration-workflows-with-recorded-logs.ts

```diff
@@ -0,0 +1,388 @@
+    await workflow.condition(() => handlerMayReturn);
+    return 'update-result';
+  });
+
+  workflow.setHandler(unfinishedHandlersWorkflowTerminationTypeSignal, async () => {
+    await workflow.condition(() => handlerMayReturn);
+  });
+
+  switch (workflowTerminationType) {
+    case 'cancellation':
+      await workflow.condition(() => false);
+    case 'continue-as-new':
+      if (waitAllHandlersFinished) {
+        handlerMayReturn = true;
+        await workflow.condition(workflow.allHandlersFinished);
+      }
+      await workflow.continueAsNew('return');
+    case 'failure':
+      throw new workflow.ApplicationFailure('Deliberately failing workflow with an unfinished handler');
+    case 'return':
+      if (waitAllHandlersFinished) {
+        handlerMayReturn = true;
+        await workflow.condition(workflow.allHandlersFinished);
+      }
+  }
+}
+
+// These tests confirm that the warning is issued / not issued as appropriate for workflow
+// termination via cancellation, continue-as-new, failure, and return, and that there is no warning
+// when waiting on allHandlersFinished before return or continue-as-new.
+test('unfinished update handler with workflow return', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'update', 'return').testWarningIsIssued(true);
+});
+
+test('unfinished update handler with workflow return waiting for all handlers to finish', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(
+    t,
+    'update',
+    'return',
+    'wait-all-handlers-finished'
+  ).testWarningIsIssued(false);
+});
+
+test('unfinished signal handler with workflow return', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'signal', 'return').testWarningIsIssued(true);
+});
+
+test('unfinished signal handler with workflow return waiting for all handlers to finish', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(
+    t,
+    'signal',
+    'return',
+    'wait-all-handlers-finished'
+  ).testWarningIsIssued(false);
+});
+
+test('unfinished update handler with workflow cancellation', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'update', 'cancellation').testWarningIsIssued(false);
+});
+
+test('unfinished signal handler with workflow cancellation', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'signal', 'cancellation').testWarningIsIssued(false);
+});
+
+test('unfinished update handler with continue-as-new', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'update', 'continue-as-new').testWarningIsIssued(true);
+});
+
+test('unfinished update handler with continue-as-new waiting for all handlers to finish', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(
+    t,
+    'update',
+    'continue-as-new',
+    'wait-all-handlers-finished'
+  ).testWarningIsIssued(false);
+});
```

Comment:
This test is flaky. Fixing that is the last thing needed in this PR.

Reply:
I have papered over that test failure in this PR; I believe it is a server bug. Here's a full description (no issue yet):

--

I have a workflow that CANs in the same WFT as an update is completed.
I ensure that the update is sent in the first WFT and, most of the time, I see what I expect:

1. worker receives WFT: `[doUpdate, startWorkflow]`
2. worker sends WFT completion: `[accept&completeUpdate, CAN]`
3. update caller gets successful update response
4. worker receives WFT after CAN: `[startWorkflow]`
5. workflow completes

However, sometimes (~1/20) I see

1. \<as above\>
2. \<as above>
3. server error when handling UpdateWorkflowExecution: [`unable to locate current workflow execution`](https://github.com/temporalio/temporal/blob/99f15c4dabf9a22763aaa13653c83a832bd0fffa/service/history/statemachine_environment.go#L167) (update caller does not get a response)
4. WFT after CAN contains the update for a second time, in addition to startWorkflow

I assume what's happening is that the server error causes an internal retry (FE => History) of the UpdateWorkflowExecution, and this retry lands on the 2nd run, after CAN. But I haven't yet dug into why the server error happens
```
	// for close workflow we need to check if it is still the current run
	// since it's possible that the workflowID has a newer run before it's locked 

wcache.GetCurrentRunID(...) != wfContext.GetWorkflowKey().RunID
```

Reply:
Opened https://github.com/temporalio/temporal/issues/6375